ext/git.c

```diff
@@ -0,0 +1,231 @@
+    snprintf(current_dir, sizeof(current_dir), "%s", start_dir);
+
+    while (true) {
+        char git_dir[PATH_MAX];
+        snprintf(git_dir, sizeof(git_dir), "%s/.git", current_dir);
+
+        if (access(git_dir, F_OK) == 0) {
+            return zend_string_init(git_dir, strlen(git_dir), 0);
+        }
+
+        char *last_slash = strrchr(current_dir, '/');
+        if (!last_slash) {
+            return NULL;
+        }
+
+        *last_slash = '\0';
+    }
+}
+
+void remove_credentials(zend_string *repo_url) {
+    char *url = ZSTR_VAL(repo_url);
+    char *at = strchr(url, '@');
+    if (at != NULL) {
+        char *start = strstr(url, "://");
+        if (start != NULL) {
+            start += 3; // Move pointer past "://"
+            size_t remaining_length = strlen(at + 1);
+            memmove(start, at + 1, remaining_length + 1);
+            ZSTR_LEN(repo_url) = (start - url) + remaining_length;
+        }
+    }
+}
+
+bool add_git_info(zval *carrier, zend_string *commit_sha, zend_string *repository_url) {
+    if (commit_sha && repository_url && ZSTR_LEN(commit_sha) > 0 && ZSTR_LEN(repository_url) > 0) {
+        remove_credentials(repository_url);
+        object_init_ex(carrier, ddtrace_ce_git_metadata);
+        ddtrace_git_metadata *git_meta = (ddtrace_git_metadata *) Z_OBJ_P(carrier);
+        ZVAL_STR(&git_meta->property_commit, commit_sha);
+        ZVAL_STR(&git_meta->property_repository, repository_url);
+        return true;
+    }
+
+    return false;
+}
+
+bool inject_from_env(zval *carrier) {
+    zend_string *commit_sha = get_DD_GIT_COMMIT_SHA();
+    zend_string *repository_url = get_DD_GIT_REPOSITORY_URL();
+    return add_git_info(carrier, commit_sha, repository_url);
+}
+
+bool inject_from_global_tags(zval *carrier) {
+    zend_array *global_tags = get_DD_TAGS();
+    zval *commit_sha = zend_hash_str_find(global_tags, ZEND_STRL("git.commit.sha"));
+    zval *repository_url = zend_hash_str_find(global_tags, ZEND_STRL("git.repository_url"));
+    if (commit_sha && repository_url && Z_TYPE_P(commit_sha) == IS_STRING && Z_TYPE_P(repository_url) == IS_STRING) {
+        return add_git_info(carrier, Z_STR_P(commit_sha), Z_STR_P(repository_url));
+    }
+    return false;
+}
+
+bool inject_from_git_dir(zval *carrier, zend_string *cwd) {
+    zend_string *git_dir = find_git_dir(ZSTR_VAL(cwd));
+    if (!git_dir) {
+        return false;
+    }
+
+    zend_string *commit_sha = get_commit_sha(ZSTR_VAL(git_dir));
+    zend_string *repository_url = get_repository_url(ZSTR_VAL(git_dir));
+    bool success = add_git_info(carrier, commit_sha, repository_url);
+    zend_string_release(git_dir);
+
+    return success;
+}
+
+void ddtrace_inject_git_metadata(zval *git_metadata_zv) {
+    zend_string *cwd_zstr = NULL;
+
+    if (SG(options) & SAPI_OPTION_NO_CHDIR) {
+        const char *script_filename = SG(request_info).path_translated;
+        const char *last_slash = strrchr(script_filename, '/');
```

Comment:
You'll have to handle `\\` (in _addition_ to `/`) `#ifdef _WIN32`.

Reply:
_in addition_, you may also see / in paths on Windows. Like both are valid separators. As far as I gather from the code, it's not normalized for you.